packages/react-components/src/components/data-providers/tsp-data-provider.ts

```diff
@@ -60,48 +56,38 @@ export class TspDataProvider {
             };
         }
 
+        // Fire all TSP requests
         this.totalRange = totalTimeRange.getEnd() - totalTimeRange.getStart();
-
         const start = totalTimeRange.getStart() + viewRange.start;
         const end = totalTimeRange.getStart() + viewRange.end;
         const timeGraphStateParams = QueryHelper.selectionTimeRangeQuery(start, end, nbTimes, ids);
-        const tspClientResponse = await this.client.fetchTimeGraphStates(this.traceUUID, this.outputId, timeGraphStateParams);
-        const stateResponse = tspClientResponse.getModel();
-        if (tspClientResponse.isOk() && stateResponse) {
-            this.timeGraphRows = stateResponse.model.rows;
-            this.timeGraphRowsOrdering(ids);
-        } else {
-            this.timeGraphRows = [];
-        }
-
-        // the start time which is normalized to logical 0 in timeline chart.
-        const chartStart = totalTimeRange.getStart();
-        const rows: TimelineChart.TimeGraphRowModel[] = [];
-        this.timeGraphRows.forEach((row: TimeGraphRow) => {
-            const rowId: number = row.entryId;
-            const entry = this.timeGraphEntries.find(tgEntry => tgEntry.id === rowId);
-            if (entry) {
-                rows.push(this.getRowModel(row, chartStart, rowId, entry));
-            }
-        });
+        const statesPromise = this.client.fetchTimeGraphStates(this.traceUUID, this.outputId, timeGraphStateParams);
 
         const additionalProps: { [key: string]: any } = {};
-
         if (annotationMarkers) {
             additionalProps['requested_marker_categories'] = annotationMarkers;
         }
         if (markerSetId) {
             additionalProps['requested_marker_set'] = markerSetId;
         }
-
         const annotationParams = QueryHelper.selectionTimeRangeQuery(start, end, nbTimes, ids, additionalProps);
-
         const annotations: Map<number, TimelineChart.TimeGraphAnnotation[]> = new Map();
-        const tspClientResponse2 = await this.client.fetchAnnotations(this.traceUUID, this.outputId, annotationParams);
-        const annotationsResponse = tspClientResponse2.getModel();
-        const rangeEvents: TimelineChart.TimeGraphAnnotation[] = [];
+        const annotationsPromise = this.client.fetchAnnotations(this.traceUUID, this.outputId, annotationParams);
+
+        const arrowStart = viewRange.start + this.timeGraphEntries[0].start;
+        const arrowEnd = viewRange.end + this.timeGraphEntries[0].start;
+        const fetchParameters = QueryHelper.timeRangeQuery(arrowStart, arrowEnd, nbTimes);
+        const arrowsPromise = this.client.fetchTimeGraphArrows(this.traceUUID, this.outputId, fetchParameters);
+
+        // Wait for responses
+        const [tspClientAnnotationsResponse, tspClientStatesResponse, tspClientArrowsResponse] = [await annotationsPromise, await statesPromise, await arrowsPromise];
```

Comment:
Use the following Promise.all() to make sure that the 3 promises are triggered at the same time:
`const [tspClientAnnotationsResponse, tspClientStatesResponse, tspClientArrowsResponse] = await Promise.all([annotationsPromise, statesPromise, arrowsPromise]);`